### ISSUE TITLE
Warn about semantics predicate at the beginning of lexer rule because of it degradates performance a lot

### ISSUE THREAD

**Issue by KvanTTT** [antlr/antlr4#3633]
From the article: https://tomassetti.me/improving-the-performance-of-an-antlr-parser/ :

> Mike Lischke, pointed out is that you should not put one at the beginning of a lexer rule. That is because rules with semantic predicates at the beginning are not cached. And in lexer rules the effect of no caching, is really heavy. **All lexer rules are handled as if they were alternatives in a single big outer rule.** So, even if only one lexer rule contains a predicate on the left hand side, no lexer rule is cached at all! When you consider that all your keywords also are lexer rules, you can easily imagine how thousands of lexer rules behave, performance-wise, if they are not cached.

I've also encountered the degradation several times.

```g4
TOKEN_1: {code}? 'TOKEN_1; // Warning
TOKEN_2: 'TOKEN_2' {code}?; // Ok
```

@mike-lischke what do you think about introducing of such warning?

**Comment by mike-lischke** [comment]
You mean having the ANTLR4 tool spit out a warning like that?

**Comment by KvanTTT** [comment]
Yes, something like `Semantics predicate at the beginning of the lexer rule significantly degradates performance`

**Comment by mike-lischke** [comment]
Sounds good, sure!

**Comment by KvanTTT** [comment]
Also, I suggest moving this info from the article to the ANTLR documentation.

**Comment by KvanTTT** [comment]
After ATN invesetigating I'm not sure it works fine and this warning should exist at all. I don't think it's correct that predicate is being called on every rule, it should be called once.
